includes/Core/Util/WP_Context_Switcher.php

```diff
@@ -0,0 +1,124 @@
+<?php
+/**
+ * Class Google\Site_Kit\Core\Util\WP_Context_Switcher
+ *
+ * @package   Google\Site_Kit\Core\Util
+ * @copyright 2020 Google LLC
+ * @license   https://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
+ * @link      https://sitekit.withgoogle.com
+ */
+
+namespace Google\Site_Kit\Core\Util;
+
+/**
+ * Class for temporarily switching WordPress context, e.g. from admin to frontend.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+class WP_Context_Switcher {
+
+	/**
+	 * Stack of original contexts switched from.
+	 *
+	 * @since n.e.x.t
+	 * @var array
+	 */
+	private static $context_stack = array();
+
+	/**
+	 * Switches to WordPress frontend context if necessary.
+	 *
+	 * Context is only switched if WordPress is not already in frontend context. Context should only ever be switched
+	 * temporarily. Call the returned closure as soon as possible after to restore the original context.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable Closure that restores context and returns true if context was restored or false otherwise.
+	 */
+	public static function with_frontend_context() {
+		global $current_screen;
+
+		if ( ! is_admin() ) {
+			return function() {
+				return false;
+			};
+		}
+
+		self::switch_current_screen( 'front' );
+		return self::get_restore_closure();
+	}
+
+	/**
+	 * Switches to WordPress admin context if necessary.
+	 *
+	 * Context is only switched if WordPress is not already in admin context. Context should only ever be switched
+	 * temporarily. Call the returned closure as soon as possible after to restore the original context.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable Closure that restores context and returns true if context was restored or false otherwise.
+	 */
+	public static function with_admin_context() {
+		global $current_screen;
+
+		if ( is_admin() ) {
+			return function() {
+				return false;
+			};
+		}
+
+		self::switch_current_screen( 'index' );
+		return self::get_restore_closure();
+	}
+
+	/**
+	 * Switches the current WordPress screen via the given screen ID or hook name.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $screen_id WordPress screen ID.
+	 */
+	private static function switch_current_screen( $screen_id ) {
+		global $current_screen;
+
+		self::load_wp_screen_api();
+		self::$context_stack[] = $current_screen;
+		$current_screen        = \WP_Screen::get( $screen_id ); // phpcs:ignore WordPress.WP.GlobalVariablesOverride.Prohibited
+	}
+
+	/**
+	 * Gets closure to restore previous WordPress context if it has been switched.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return callable Closure that restores context and returns true if context was restored or false otherwise.
+	 */
+	private static function get_restore_closure() {
+		return static function() {
+			global $current_screen;
+
+			if ( empty( self::$context_stack ) ) { // phpcs:ignore WordPressVIPMinimum.Variables.VariableAnalysis.SelfInsideClosure
```

Comment:
I thought about possibly using a stack for this but I don't think there is a need for it. It shouldn't really matter in our current use as I don't think we have any nested context switching, but I think restoring in this way is riskier than simply restoring the previous screen as I originally suggested.

Re-setting the previous screen is an idempotent operation which guarantees that things are in fact back the way they were after the restore is called. Using a stack, every restore function is the same and is capable of restoring layers that it didn't "push".

E.g.
```php
$restore();
$restore();
$restore();
```

This is a contrived example of course, but just showing that this approach could introduce problems that wouldn't exist with the other. My [suggested approach](https://github.com/google/site-kit-wp/issues/1978#issuecomment-684014717) still supports nesting if indeed needed and is still stack-based of course (using the call stack), but is a bit more deterministic.